services/publishing/Tweek.Publishing.Service/Utils/ShellHelper.cs

```diff
@@ -0,0 +1,138 @@
+                    return (read, buffer);
+                })
+                .Repeat()
+                .TakeWhile(x => x.read > 0)
+                .Select(x => x.read == bufferSize ? x.buffer : x.buffer.Slice(x.read));
+        }
+    }
+
+    public static class ShellHelper
+    {
+        public delegate (Process, Task) ShellExecutor(string command, string args, Action<ProcessStartInfo> paramsInit = null);
+
+        public static readonly ShellExecutor Executor = ExecProcess;
+
+        public static (Process, Task) ExecProcess(string command, string args, Action<ProcessStartInfo> paramsInit = null)
+        {
+            var startInfo = new ProcessStartInfo
+            {
+                FileName = command,
+                Arguments = args,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false,
+                CreateNoWindow = true
+            };
+
+            paramsInit?.Invoke(startInfo);
+
+            var process = new Process
+            {
+                EnableRaisingEvents = true,
+                StartInfo = startInfo,
+            };
+
+            var exited = Observable.FromEventPattern(x => process.Exited += x, x => process.Exited -= x).Take(1).ToTask();
+
+            if (!process.Start())
+            {
+                throw new Exception("failed to start process");
+            }
+
+            return (process, exited);
+        }
+
+        public static IObservable<(byte[] data, OutputType outputType)> ExecObservable(this ShellExecutor shellExecutor,
+            string command, string args, Action<ProcessStartInfo> paramsInit = null)
+        {
+            return Observable.Defer(() =>
+            {
+                var (process, exited) = shellExecutor(command, args, paramsInit);
+                var sbErr = new StringBuilder();
+
+                return process.StandardOutput.BaseStream.ToObservable().Select(data => (data, OutputType.StdOut))
+                    .Merge(process.StandardError.BaseStream.ToObservable()
+                        .Do(data => sbErr.Append(Encoding.Default.GetString(data)))
+                        .Select(data => (data, OutputType.StdErr))
+                    )
+                    .With(exited.ToObservable().Do(_ =>
+                    {
+                        if (process.ExitCode != 0)
+                            throw new Exception("proccess failed")
+                            {
+                                Data =
+                                {
+                                    ["Command"] = command,
+                                    ["Args"] = args,
+                                    ["ExitCode"] = process.ExitCode,
+                                    ["StdErr"] = sbErr.ToString()
+                                }
+                            };
+                    }));
+            }).Publish().RefCount();
```

Comment:
I don't think that a method that returns an observable should be the one that does `.Publish().RefCount()`.
If whoever called this method expects it to be shared, he should be the one to do it.

and it looks like the places that call this method don't really need this observable to be shared anyway

Reply:
Removed